networkx/algorithms/approximation/traveling_salesman.py

```diff
@@ -321,6 +333,762 @@ def threshold_tsp(G, weight):
+                # Check that delta had the desired effect
+                new_q_e = q(e)
+                desired_q_e = (1 + EPSILON / 2) * z_e
+                if round(new_q_e, 8) != round(desired_q_e, 8):
+                    raise Exception
+            else:
+                in_range_count += 1
+        # Check if the for loop terminated without changing any gamma
+        if in_range_count == len(gamma):
+            break
+
+    # Remove the new edge attributes
+    for _, _, d in G.edges(data=True):
+        if lambda_key in d:
+            del d[lambda_key]
+
+    return gamma
+
+
+@py_random_state(2)
+def sample_spanning_tree(G, lambda_key, seed=None):
+    """
+    Sample a spanning tree using the edges weights of the graph.
+
+    The edge weights are multiplicative, so the probability of each tree is
+    proportional to the product of edge weights.
+
+    The algorithm itself uses algorithm A8 in [1]_ .
+
+    We 'shuffle' the edges in the graph, and then probabilistically
+    determine weather to add the edge conditioned on all of the previous
+    edges which where added to the tree. Probabilities are calculated using
+    Kirchhoff's Matrix Tree Theorem and a weighted Laplacian matrix.
+
+    At each iteration, we contract the edges we have decided to include in the
+    sampled tree and delete those which we have decided not to include.
+
+    Parameters
+    ----------
+    G : nx.Graph
+        An undirected version of the original graph.
+
+    lambda_key : string
+        The edge key for the edge attribute holding edge weight.
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    nx.Graph
+        A spanning tree using the distribution defined by `gamma`.
+
+    References
+    ----------
+    .. [1] V. Kulkarni, Generating random combinatorial objects, Journal of
+       algorithms, 11 (1990), pp. 185–207
+    """
+
+    def find_node(merged_nodes, n):
+        """
+        We can think of clusters of contracted nodes as having one
+        representative in the graph. Each node which is not in merged_nodes
+        is still its own representative. Since a representative can be later
+        contracted, we need to recursively search though the dict to find
+        the final representative, but once we know it we can use path
+        compression to speed up the access of the representative for next time.
+
+        Parameters
+        ----------
+        merged_nodes : dict
+            The dict storing the mapping from node to representative
+        n
+            The node whose representative we seek
+
+        Returns
+        -------
+        The representative of the `n`
+        """
+        if n not in merged_nodes:
+            return n
+        else:
+            rep = find_node(merged_nodes, merged_nodes[n])
+            merged_nodes[n] = rep
+            return rep
```

Comment:
Another minor suggestion that is getting into the woods somewhat.  This `find_node` function is recursive and python recursion can be tricky for users if 1) the recursion limit has to be reset, or 2) the overhead of all the function calls becomes large.  Sometimes the recursive code is more readable but many times not. All recursive functions can be "unrolled" to be a loop which uses a stack to store the information from one function namespace so it can be used on a subsequent loop. The loop iterations act as the function calls with variables being put onto the stack instead of "calling" and taken off the stack instead of "returning". In this case the non-recursive function is something like this:

```python
    rep = n
    traced_nodes = []
    while rep in merged_nodes:
        traced_nodes.append(rep)
        rep = merged_nodes[rep]
    # compress path to speed up future calls
    for node in traced_nodes:
        merged_nodes[node] = rep
    return rep
```

I'll leave it up to you as to which is more readable and whether the recursion limit might be reached with `find_node`.

Reply:
I was not aware of some of the python quirks with regard to recursion. But, the default python recursion limit is 1000. Ordinarily this is no insurance that the structure itself will not reach that limit, but it is important to remember that this is part of the atsp algorithm and in order to reach the limit even under worst case scenarios you would need a graph with more than 1000 nodes in it. I believe that because the large graph size, running `asadpour_atsp` would not be feasible due to performance limits, notably with respect to the Held Karp relaxation.